core/indigo-core/reaction/src/reaction_multistep_detector.cpp

```diff
@@ -238,6 +227,430 @@ void ReactionMultistepDetector::createSummBlocks()
+            if (cont > max_containment)
+            {
+                max_containment = cont;
+                result = {i, j};
+            }
+            if (cont > 0.0f)
+            {
+                auto& zone_it = zones.find(i);
+                if (zone_it != zones.end())
+                    zone_it->second.insert(j);
+                else
+                    zones.insert({i, {j}});
+            }
+        }
+    }
+    return result;
+}
+
+void ReactionMultistepDetector::mergeCloseComponents()
+{
+    for (std::size_t i = 0; i < _components.size(); ++i)
+    {
+        if (!_components[i].first)
+            continue;
+        std::queue<std::pair<std::size_t, std::optional<std::pair<int, int>>>> bfs_queue;
+        std::vector<std::size_t> cluster;
+        bfs_queue.emplace(i, std::nullopt);
+        cluster.push_back(i);
+        while (!bfs_queue.empty())
+        {
+            auto qel = bfs_queue.front();
+            bfs_queue.pop();
+            for (std::size_t j = 0; j < _components.size(); ++j)
+            {
+                if (!_components[j].first || j == qel.first)
+                    continue;
+                if (std::find(cluster.begin(), cluster.end(), j) != cluster.end())
+                    continue;
+                auto zone = isMergeable(qel.first, j, qel.second);
+                if (zone.has_value())
+                {
+                    cluster.push_back(j);
+                    bfs_queue.emplace(j, zone);
+                }
+            }
+        }
+        for (std::size_t k = 1; k < cluster.size(); ++k)
+        {
+            QS_DEF(Array<int>, mapping);
+            if (_components[cluster[k]].first)
+            {
+                _components[i].first->mergeWithMolecule(*_components[cluster[k]].first, &mapping, 0);
+                _components[cluster[k]].first.reset();
+            }
+        }
+    }
+    _components.erase(std::remove_if(_components.begin(), _components.end(), [](auto& p) { return !p.first; }), _components.end());
+    _moleculeCount = (int)_components.size();
+}
+
+std::optional<std::pair<int, int>> ReactionMultistepDetector::isMergeable(size_t mol_idx1, size_t mol_idx2, std::optional<std::pair<int, int>> current_zone)
+{
+    auto& mdi = _mol_distances[mol_idx1];
+    auto dist_it = mdi.distances_map.find(mol_idx2);
+    if (dist_it != mdi.distances_map.end() && dist_it->second < LayoutOptions::DEFAULT_BOND_LENGTH * 2)
+    {
+        // collect surrounding zones for both molecules
+        std::map<int, std::unordered_set<int>> other_zones1;
+        std::map<int, std::unordered_set<int>> other_zones2;
+        std::map<int, std::unordered_set<int>> comm_zones;
+
+        auto zone1 = findMaxSpecialZone(mol_idx1, other_zones1);
+        auto zone2 = findMaxSpecialZone(mol_idx2, other_zones2);
+        std::set_intersection(other_zones1.begin(), other_zones1.end(), other_zones2.begin(), other_zones2.end(), std::inserter(comm_zones, comm_zones.begin()),
```

Comment:
I think intersection calc better move to line 573

Reply:
Now it is line 595 - just before `for (auto& cz : comm_zones)`
No need to calculate intersection before this point.

Reply:
fixed